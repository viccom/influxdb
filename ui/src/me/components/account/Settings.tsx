// Libraries
import React, {PureComponent, ChangeEvent} from 'react'
import {connect} from 'react-redux'

// Types
import {MeState} from 'src/types/v2'
import {
  Form,
  Button,
  Input,
  ComponentSize,
  ComponentStatus,
  Panel,
  Grid,
  Columns,
} from 'src/clockface'

interface StateProps {
  me: MeState
}

interface State {
  me: MeState
}

export class Settings extends PureComponent<StateProps, State> {
  constructor(props) {
    super(props)
    this.state = {
      me: this.props.me,
    }
  }

  public render() {
    const {me} = this.state

    return (
      <Grid>
        <Grid.Row>
          <Grid.Column widthXS={Columns.Six}>
            <Panel>
              <Panel.Header title="About Me">
                <Button text="Edit About Me" />
              </Panel.Header>
              <Panel.Body>
                <Form>
                  <Form.Element label="Username">
                    <Input
                      value={me.name}
                      dataTest="nameInput"
                      titleText="Username"
                      size={ComponentSize.Small}
                      status={ComponentStatus.Disabled}
                      onChange={this.handleChangeInput}
                    />
                  </Form.Element>
                </Form>
              </Panel.Body>
            </Panel>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    )
  }

  private handleChangeInput = (_: ChangeEvent<HTMLInputElement>): void => {
    //  console.log('changing: ', e)
  }
}

const mstp = ({me}) => ({
  me,
})

export default connect<StateProps>(mstp)(Settings)
