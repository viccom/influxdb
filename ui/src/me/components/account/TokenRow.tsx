// Libraries
import React, {PureComponent} from 'react'

// Components
import {
  IndexList,
  ComponentSpacer,
  Alignment,
  Button,
  ComponentSize,
  ComponentColor,
} from 'src/clockface'

// Types
import {Authorization} from 'src/api'

interface Props {
  auth: Authorization
  onClickDescription: (authID: string) => void
}

export default class TokenRow extends PureComponent<Props> {
  public render() {
    const {description, status, id} = this.props.auth

    return (
      <IndexList.Row>
        <IndexList.Cell>
          <a
            href="#"
            onClick={this.handleClickDescription}
            data-test={`token-description-${id}`}
          >
            {description}
          </a>
        </IndexList.Cell>
        <IndexList.Cell>{status}</IndexList.Cell>
        <IndexList.Cell alignment={Alignment.Right} revealOnHover={true}>
          <ComponentSpacer align={Alignment.Right}>
            <Button
              size={ComponentSize.ExtraSmall}
              color={ComponentColor.Danger}
              text="Delete"
            />
          </ComponentSpacer>
        </IndexList.Cell>
      </IndexList.Row>
    )
  }

  private handleClickDescription = () => {
    const {onClickDescription, auth} = this.props
    onClickDescription(auth.id)
  }
}
