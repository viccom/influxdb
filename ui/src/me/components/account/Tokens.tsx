// Libraries
import React, {PureComponent, ChangeEvent} from 'react'
import {connect} from 'react-redux'

// Components
import {IconFont, Input, Spinner} from 'src/clockface'
import ResourceFetcher from 'src/shared/components/resource_fetcher'
import TokenList from 'src/me/components/account/TokensList'
import FilterList from 'src/shared/components/Filter'
import TabbedPageHeader from 'src/shared/components/tabbed_page/TabbedPageHeader'

// APIs
import {getAuthorizations} from 'src/authorizations/apis'

// Actions
import {notify} from 'src/shared/actions/notifications'
import {NotificationAction} from 'src/types'

// Types
import {Authorization} from 'src/api'

interface State {
  searchTerm: string
}

enum AuthSearchKeys {
  Description = 'description',
  Status = 'status',
}

interface Props {
  onNotify: NotificationAction
}

export class Tokens extends PureComponent<Props, State> {
  constructor(props) {
    super(props)
    this.state = {
      searchTerm: '',
    }
  }

  public render() {
    const {onNotify} = this.props
    const {searchTerm} = this.state

    return (
      <>
        <TabbedPageHeader>
          <Input
            icon={IconFont.Search}
            value={searchTerm}
            placeholder="Filter Tokens..."
            onChange={this.handleChangeSearchTerm}
            widthPixels={256}
          />
        </TabbedPageHeader>
        <ResourceFetcher<Authorization[]> fetcher={getAuthorizations}>
          {(fetchedAuths, loading) => (
            <Spinner loading={loading}>
              <FilterList<Authorization>
                list={fetchedAuths}
                searchTerm={searchTerm}
                searchKeys={this.searchKeys}
              >
                {filteredAuths => (
                  <TokenList
                    auths={filteredAuths}
                    onNotify={onNotify}
                    searchTerm={searchTerm}
                  />
                )}
              </FilterList>
            </Spinner>
          )}
        </ResourceFetcher>
      </>
    )
  }

  private handleChangeSearchTerm = (e: ChangeEvent<HTMLInputElement>): void => {
    this.setState({searchTerm: e.target.value})
  }

  private get searchKeys(): AuthSearchKeys[] {
    return [AuthSearchKeys.Status, AuthSearchKeys.Description]
  }
}

const mdtp = {
  onNotify: notify,
}

export default connect<Props>(
  null,
  mdtp
)(Tokens)
