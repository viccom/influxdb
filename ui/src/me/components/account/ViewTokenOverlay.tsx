// Libraries
import React, {PureComponent} from 'react'

// Components
import {OverlayContainer, OverlayBody, OverlayHeading} from 'src/clockface'
import PermissionsWidget, {
  PermissionsWidgetMode,
  PermissionsWidgetSelection,
} from 'src/shared/components/permissionsWidget/PermissionsWidget'
import CopyText from 'src/shared/components/CopyText'

// Types
import {Authorization, Permission} from 'src/api'

// Actions
import {NotificationAction} from 'src/types'

const {Write, Read} = Permission.ActionEnum

interface Props {
  onNotify: NotificationAction
  auth: Authorization
  onDismissOverlay: () => void
}

const actions = [Read, Write]

export default class ViewTokenOverlay extends PureComponent<Props> {
  public render() {
    const {description, permissions} = this.props.auth
    const {onNotify} = this.props

    return (
      <OverlayContainer>
        <OverlayHeading title={description} onDismiss={this.handleDismiss} />
        <OverlayBody>
          <CopyText copyText={this.props.auth.token} notify={onNotify} />
          <PermissionsWidget
            mode={PermissionsWidgetMode.Read}
            heightPixels={500}
          >
            {permissions.map((p, i) => {
              return (
                <PermissionsWidget.Section
                  key={i}
                  id={this.id(p)}
                  title={this.title(p)}
                  mode={PermissionsWidgetMode.Read}
                >
                  {actions.map((a, i) => (
                    <PermissionsWidget.Item
                      key={i}
                      id={this.itemID(p, a)}
                      label={a}
                      selected={this.selected(p, a)}
                    />
                  ))}
                </PermissionsWidget.Section>
              )
            })}
          </PermissionsWidget>
        </OverlayBody>
      </OverlayContainer>
    )
  }

  private selected = (
    permission: Permission,
    action: Permission.ActionEnum
  ): PermissionsWidgetSelection => {
    if (permission.action === action) {
      return PermissionsWidgetSelection.Selected
    }

    return PermissionsWidgetSelection.Unselected
  }

  private itemID = (
    permission: Permission,
    action: Permission.ActionEnum
  ): string => {
    return `${permission.resource.type}-${action}-${permission.resource.id ||
      '*'}-${permission.resource.orgID || '*'}`
  }

  private id = (permission: Permission): string => {
    return permission.resource.type
  }

  private title = (permission: Permission): string => {
    const org = permission.resource.org || '*'
    const name = permission.resource.name || '*'
    const type = permission.resource.type

    if (permission.resource.name || permission.resource.org) {
      return `${org}:${type}:${name}`
    }

    return `${permission.resource.type}:*`
  }

  private handleDismiss = () => {
    this.props.onDismissOverlay()
  }
}
