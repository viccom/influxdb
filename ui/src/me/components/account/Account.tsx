// Libraries
import React, {PureComponent} from 'react'
import {WithRouterProps} from 'react-router'

// Components
import {Page} from 'src/pageLayout'
import TabbedPage from 'src/shared/components/tabbed_page/TabbedPage'
import TabbedPageSection from 'src/shared/components/tabbed_page/TabbedPageSection'
import Settings from 'src/me/components/account/Settings'
import Tokens from 'src/me/components/account/Tokens'

export enum Tabs {
  Settings = 'settings',
  Tokens = 'tokens',
}

export default class Account extends PureComponent<WithRouterProps> {
  public render() {
    const {params} = this.props
    return (
      <Page titleTag="Account">
        <Page.Header fullWidth={false}>
          <Page.Header.Left>
            <Page.Title title="Account" />
          </Page.Header.Left>
          <Page.Header.Right />
        </Page.Header>
        <Page.Contents fullWidth={false} scrollable={true}>
          <TabbedPage
            name="Account"
            activeTabUrl={params.tab}
            parentUrl="/account"
          >
            <TabbedPageSection
              title="My Settings"
              id={Tabs.Settings}
              url={Tabs.Settings}
            >
              <Settings />
            </TabbedPageSection>
            <TabbedPageSection
              title="My Tokens"
              id={Tabs.Tokens}
              url={Tabs.Tokens}
            >
              <Tokens />
            </TabbedPageSection>
          </TabbedPage>
        </Page.Contents>
      </Page>
    )
  }
}
