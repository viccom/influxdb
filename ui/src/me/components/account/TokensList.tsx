// Libraries
import React, {PureComponent} from 'react'

// Components
import {
  IndexList,
  EmptyState,
  ComponentSize,
  OverlayTechnology,
} from 'src/clockface'
import TokenRow from 'src/me/components/account/TokenRow'

// Types
import {Authorization} from 'src/api'
import ViewTokenOverlay from './ViewTokenOverlay'

// Actions
import {NotificationAction} from 'src/types'

interface Props {
  auths: Authorization[]
  onNotify: NotificationAction
  searchTerm: string
}

interface State {
  isTokenOverlayVisible: boolean
  authInView: Authorization
}

export default class TokenList extends PureComponent<Props, State> {
  constructor(props) {
    super(props)
    this.state = {
      isTokenOverlayVisible: false,
      authInView: null,
    }
  }

  public render() {
    const {auths, onNotify} = this.props
    const {isTokenOverlayVisible, authInView} = this.state

    return (
      <>
        <IndexList>
          <IndexList.Header>
            <IndexList.HeaderCell columnName="Description" />
            <IndexList.HeaderCell columnName="Status" />
          </IndexList.Header>
          <IndexList.Body emptyState={this.emptyState} columnCount={2}>
            {auths.map(a => {
              return (
                <TokenRow
                  key={a.id}
                  auth={a}
                  onClickDescription={this.handleClickDescription}
                />
              )
            })}
          </IndexList.Body>
        </IndexList>
        <OverlayTechnology visible={isTokenOverlayVisible}>
          <ViewTokenOverlay
            onNotify={onNotify}
            auth={authInView}
            onDismissOverlay={this.handleDismissOverlay}
          />
        </OverlayTechnology>
      </>
    )
  }

  private handleDismissOverlay = () => {
    this.setState({isTokenOverlayVisible: false})
  }

  private handleClickDescription = (authID: string): void => {
    const authInView = this.props.auths.find(a => a.id === authID)
    this.setState({isTokenOverlayVisible: true, authInView})
  }

  private get emptyState(): JSX.Element {
    const {searchTerm} = this.props
    let emptyStateText =
      'There are not any Tokens associated with this account. Contact your administrator'

    if (searchTerm) {
      emptyStateText = 'No Tokens match your search term'
    }

    return (
      <EmptyState size={ComponentSize.Large}>
        <EmptyState.Text text={emptyStateText} />
      </EmptyState>
    )
  }
}
