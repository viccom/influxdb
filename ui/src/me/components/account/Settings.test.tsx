// Libraries
import React from 'react'
import {mount} from 'enzyme'

// Components
import {Settings} from 'src/me/components/account/Settings'
import {me} from '../../mockUserData'

const setup = (override?) => {
  const props = {
    me,
    ...override,
  }

  const wrapper = mount(<Settings {...props} />)

  return {wrapper}
}

describe('Account', () => {
  describe('rendering', () => {
    it('renders!', () => {
      const {wrapper} = setup()

      expect(wrapper.exists()).toBe(true)
      expect(wrapper).toMatchSnapshot()
    })

    it('displays the users info by default', () => {
      const {wrapper} = setup()

      const nameInput = wrapper.find({'data-test': 'nameInput'})
      expect(nameInput.props().value).toBe(me.name)
    })
  })
})
