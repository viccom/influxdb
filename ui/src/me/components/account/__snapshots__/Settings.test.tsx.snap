// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account rendering renders! 1`] = `
<Settings
  me={
    Object {
      "id": "id-of-groot",
      "links": Object {
        "log": "/api/v2/users/id-of-groot/log",
        "self": "/api/v2/users/id-of-groot",
      },
      "name": "groot",
    }
  }
>
  <Grid>
    <div
      className="grid--container"
    >
      <Select
        type={[Function]}
      >
        <GridRow
          key=".0"
        >
          <div
            className="grid--row"
          >
            <GridColumn
              widthXS={6}
            >
              <div
                className="grid--column col-xs-6"
              >
                <Panel>
                  <div
                    className="panel"
                  >
                    <PanelHeader
                      title="About Me"
                    >
                      <div
                        className="panel-header"
                      >
                        <div
                          className="panel-title"
                        >
                          About Me
                        </div>
                        <div
                          className="panel-controls"
                        >
                          <ComponentSpacer
                            align="right"
                          >
                            <div
                              className="component-spacer component-spacer--right component-spacer--horizontal"
                            >
                              <Button
                                active={false}
                                color="default"
                                shape="none"
                                size="sm"
                                status="default"
                                text="Edit About Me"
                                type="button"
                              >
                                <button
                                  className="button button-sm button-default"
                                  disabled={false}
                                  tabIndex={0}
                                  title="Edit About Me"
                                  type="button"
                                >
                                  Edit About Me
                                </button>
                              </Button>
                            </div>
                          </ComponentSpacer>
                        </div>
                      </div>
                    </PanelHeader>
                    <PanelBody>
                      <div
                        className="panel-body"
                      >
                        <Form>
                          <form
                            className="form--wrapper"
                            onSubmit={[Function]}
                          >
                            <FormElement
                              label="Username"
                            >
                              <div
                                className="form--element"
                              >
                                <FormLabel
                                  label="Username"
                                >
                                  <label
                                    className="form--label"
                                  >
                                    <span>
                                      Username
                                    </span>
                                  </label>
                                </FormLabel>
                                <Input
                                  autoFocus={false}
                                  autocomplete="off"
                                  dataTest="nameInput"
                                  disabledTitleText="This input is disabled"
                                  name=""
                                  onChange={[Function]}
                                  placeholder=""
                                  size="sm"
                                  spellCheck={false}
                                  status="disabled"
                                  titleText="Username"
                                  value="groot"
                                >
                                  <div
                                    className="input input-sm input--disabled"
                                    style={
                                      Object {
                                        "width": "100%",
                                      }
                                    }
                                  >
                                    <input
                                      autoComplete="off"
                                      autoFocus={false}
                                      className="input-field"
                                      data-test="nameInput"
                                      disabled={true}
                                      name=""
                                      onChange={[Function]}
                                      placeholder=""
                                      spellCheck={false}
                                      title="This input is disabled"
                                      value="groot"
                                    />
                                    <div
                                      className="input-shadow"
                                    />
                                  </div>
                                </Input>
                              </div>
                            </FormElement>
                          </form>
                        </Form>
                      </div>
                    </PanelBody>
                  </div>
                </Panel>
              </div>
            </GridColumn>
          </div>
        </GridRow>
      </Select>
    </div>
  </Grid>
</Settings>
`;
